src/realm/object-store/sync/sync_user.cpp

```diff
@@ -409,7 +434,9 @@ void SyncUser::register_session(std::shared_ptr<SyncSession> session)
 
 app::MongoClient SyncUser::mongo_client(const std::string& service_name)
 {
-    return app::MongoClient(shared_from_this(), m_sync_manager.app().lock(), service_name);
+    std::lock_guard<std::mutex> lk(m_mutex);
+    REALM_ASSERT(m_state == SyncUser::State::LoggedIn);
```

Comment:
Should we make this an exception rather than an assertion? I can imagine a situation where the developer erroneously holds on to a logged out user and tries to invoke remote mongo operations on it. It's obviously a programming error, but exceptions are much more readable and easier to react to than assertions. This is relevant for other places where we've added assertions - we should evaluate which of those can be hit through user error and which are bugs in Core/OS.